estimator/lwe_primal.py

```diff
@@ -299,7 +299,10 @@ def cost(
 
         # 1. Simulate BKZ-β
         # TODO: pick τ
-        r = simulator(d, params.n - zeta, params.q, beta, xi=xi)
+        if simulator is CN11:
+            r = simulator(d, params.n - zeta, params.q, beta, xi=xi, dual=True)
+        else:
+            r = simulator(d, params.n - zeta, params.q, beta, xi=xi)
         bkz_cost = costf(red_cost_model, beta, d)
```

Comment:
We might be able to remove this `if/else` statement, but I couldn't figure out the best way to do it.

Reply:
Can't you just do
```
try:
  f(dual=True)
except TypeError:
 f()
```
?

Reply:
How about just adding the `dual` flag to the simulator interface? I always cringe when a function needs to know which specific instantiation of some type it is given. GSA can just ignore the flag then (but maybe add a comment why). 

Another way would be to simply do the reverse and invert of the basis in both cases as part of a generic "dual_simulate" function that may be applied to any simulator. That would uselessly apply the transformations also in the GSA case, so there might be some slow down, but if it's not too bad, that might be a nice solution, since it makes it easy to have new simulators that are compatible with the primal hybrid.

Reply:
I like the `dual_simulate` approach. Will need to extract some LLL `r` shape generically for `d,k,q,xi` but that's perhaps useful more generally, too.

Reply:
Ah, the annoying thing with `dual_simulate(simulator, d, params.n - zeta, params.q, beta, xi=xi)` is that the simulators don't expect `r` but `d,k,q,xi`. So perhaps the right thing™ is to rewrite the simulator interface to take `r` as input and then we provide our own higher level functions that take `d,k,q,xi` and produce GSO norms. These can then also transparently handle the dual case, simulators don't need to know. To avoid the slow down, we could special case the GSA thing inside these higher level functions?

Reply:
Yes, I agree that makes sense. This sounds like a bit more work, so if @bencrts  doesn't wanna do this refactoring right now, we could approve this PR as is and leave this for a separate PR.

Reply:
Agreed

Reply:
I agree that the suggested changes (e.g. `dual_simulate()`) are probably a better idea. Happy to start working towards this at some point next week.

I've fixed all of the outstanding issues in the code review, and I agree that merging is a good idea (that way we get the fix into the codebase asap).